### PR TITLE
Hide scheduler workflows with new search attribute

### PR DESCRIPTION
**Pull request by dnr**
**What changed?**
- new predefined `TemporalNamespaceDivision` search attribute + schema changes
- set via standard search attributes
- query with new field in `ListWorkflowExecutionsRequest` or in query string
- all queries default to adding a "must not exist" on the new attribute, unless explicitly selected for
- this only works for ES, standard visibility ignores search attributes and the new queries have no effect

**Why?**
To hide scheduler, batch operation, other internal workflows from normal visibility operations in user namespaces

**How did you test it?**
unit tests + manual test

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: common/persistence/visibility/store/elasticsearch/query_interceptors.go [hunk 1]
```diff
@@ -86,6 +87,10 @@ func (ni *nameInterceptor) Name(name string, usage query.FieldNameUsage) (string
 		}
 	}
 
+	if fieldName == searchattribute.TemporalNamespaceDivision && usage == query.FieldNameFilter {
```
Comment (dnr) [inline common/persistence/visibility/store/elasticsearch/query_interceptors.go:90]: This feels icky. Any better ideas?

File: schema/elasticsearch/visibility/versioned/v3/upgrade.sh [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+case $ES_VERSION in
+    v6) date_type='date'       ; doc_type='/_doc' ;;
+    *)  date_type='date_nanos' ; doc_type=''      ;;
+esac
+
+new_mapping='
+{
+  "properties": {
+    "TemporalNamespaceDivision": {
```
Comment (dnr) [inline schema/elasticsearch/visibility/versioned/v3/upgrade.sh:39]: This is getting kind of annoying. What do you think about auto-registering new pre-defined on upgrade? Or the cli tool?

File: service/frontend/workflowHandler.go [hunk 3]
```diff
@@ -4546,6 +4549,17 @@ func (wh *WorkflowHandler) cleanScheduleSearchAttributes(searchAttributes *commo
 	return searchAttributes
 }
 
+// This may mutate saPtr and *saPtr
+func (wh *WorkflowHandler) addSearchAttribute(saPtr **commonpb.SearchAttributes, key string, value *commonpb.Payload) {
```
Comment (dnr) [inline service/frontend/workflowHandler.go:4553]: Is there a helper function for this already? I didn't see one. It should probably be moved somewhere

**Comment by yiminc** [comment]
Looks reasonable to me. cc @alexshtin

File: service/frontend/workflowHandler.go (hunk 3 above)
Comment (alexshtin) [reply service/frontend/workflowHandler.go:4554]: I would move it to https://github.com/temporalio/temporal/blob/master/common/searchattribute/search_attirbute.go#L73.

File: schema/elasticsearch/visibility/versioned/v3/upgrade.sh (hunk 2 above)
Comment (alexshtin) [reply schema/elasticsearch/visibility/versioned/v3/upgrade.sh:34]: We are starting to get ES schema changes more often and expectation was the opposite. Ideally, special ES CLI tool needs to be added (there is an open #2977 PR for it) which will handle all ES schema updates. I would leave it as is for now.

File: common/persistence/visibility/store/elasticsearch/query_interceptors.go (hunk 1 above)
Comment (alexshtin) [reply common/persistence/visibility/store/elasticsearch/query_interceptors.go:90]: This is how it is supposed to be used.

**Review by alexshtin — APPROVED** [review]

File: schema/elasticsearch/visibility/versioned/v3/index_template_v6.json [hunk 4]
```diff
@@ -0,0 +1,84 @@
+{
```
[the remark this answers is not in the document]
Comment (yiminc) [reply schema/elasticsearch/visibility/versioned/v3/index_template_v6.json:1]: FYI: https://github.com/temporalio/temporal/pull/3184

[the remark this answers is not in the document]
Comment (dnr) [reply schema/elasticsearch/visibility/versioned/v3/index_template_v6.json]: I removed the new `index_template_v6.json` (and just didn't update the symlink). So this should merge nicely with #3184

**Pull request merged**